### PR TITLE
[docs] Unit testing documentation

### PR DESCRIPTION
**Pull request by ouchadam**
## Type of change

- [ ] Feature
- [ ] Bugfix
- [ ] Technical
- [x] Other :

## Content

Providing context and examples around contributing unit tests to the project, describing the current conventions and best practices. 

[Rendered](https://github.com/vector-im/element-android/blob/1b3e099d7ce4d91cb6615bc65945fc6f8495507e/docs/unit_testing.md)

## Screenshots / GIFs
N/A

## Tests
N/A

## Tested devices

N/A

### CODE REVIEW

**Review by bmarty — APPROVED** [review]
This is an amazing documentation. Thanks a lot!
Having inlined Kotlin snippets also help to illustrate and understand how it works.

**Review by manuroe — APPROVED** [review]
This doc will be our reference for a long time. Thanks!

File: docs/unit_testing.md [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+    * [Setup](#setup)
+    * [Naming](#naming)
+    * [Format](#format)
+    * [Assertions](#assertions)
+    * [Constants](#constants)
+    * [Mocking](#mocking)
+    * [Fakes](#fakes)
+    * [Fixtures](#fixtures)
+* [Examples](#examples)
+    * [Simple](#extensions-used-to-streamline-the-test-setup)
```
[the remark this answers is not in the document]
Comment (ouchadam) [reply docs/unit_testing.md:17]: ah, didn't realise :sweat_smile:  will trigger and update :+1:

File: docs/unit_testing.md [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+private fun createIntCache(cacheSize: Int): Pair<CircularCache<Int>, Array<Int?>> {
+    var internalData: Array<Int?>? = null
+    val factory: (Int) -> Array<Int?> = {
+        Array<Int?>(it) { null }.also { array -> internalData = array }
+    }
+    return CircularCache(cacheSize, factory) to internalData!!
+}
+
+private fun CircularCache<Int>.putInOrder(vararg keys: Int) {
+    keys.forEach { put(it) }
```
[the remark this answers is not in the document]
Comment (ouchadam) [reply docs/unit_testing.md:264]: happy to update both :+1:

**Pull request merged**